### PR TITLE
fixes: remove unfinished shapes, remove selectable and make shapes remain inside canvas

### LINKED ISSUE #2588: Image occlusion doesn't scale when page zoomed in
Because a max-height is set on the occlusion container, zooming the page in with ctrl+mousewheel results in the image actually shrinking slightly, due to the change in margins.

If we use an explicit height, instead of capping the height relative to the viewport, then scaling works, but images do not fit the screen by default. 

```diff
diff --git a/ts/image-occlusion/review.scss b/ts/image-occlusion/review.scss
index 8e4065fd1..95da4ba5a 100644
--- a/ts/image-occlusion/review.scss
+++ b/ts/image-occlusion/review.scss
@@ -5,7 +5,8 @@
     transform: translate(-50%, 0);
     // allow for 20px margin on html element, or short windows can truncate
     // image
-    max-height: calc(95vh - 40px);
+    //max-height: calc(95vh - 40px);
+    height: 1000px;
 }
 
 #image-occlusion-container img {
```

If anyone has ideas on how we could both scale-to-fit by default (and when window is resized), but also support zooming, please let us know!

Heads up @krmanik @glutanimate

### PR DESCRIPTION
**Pull request by krmanik**
IO fixes

#2588
#2770
#2774
#2800
#2808

### CODE REVIEW

**Comment by krmanik** [comment]
Click through transparent area have following warning.
```
Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
```
I have tried this for both upper and lower canvas, but it does not solve the warning so maybe the issues is in fabric.js.
```
canvas.getContext("2d", {willReadFrequently: true});
```

**Comment by dae** [comment]
Thanks a lot for looking into these Mani! I'll give them a proper review in the next few days after 23.10.1 is out the door.

**Comment by BEST8OY** [comment]
Not sure if I'm allowed to ask for this here, sorry in advance if I am.
A shortcut for "Toggle Masks" in reviewer would be handy.
Something like "M"

**Comment by krmanik** [comment]
> Not sure if I'm allowed to ask for this here, sorry in advance if I'm not. A shortcut for "Toggle Masks" in reviewer would be handy. Something like "M"

Add key H for toggle mask because, M is already assigned to menu.

File: ts/image-occlusion/mask-editor.ts [hunk 1]
```diff
@@ -120,8 +120,12 @@ function initCanvas(onChange: () => void): fabric.Canvas {
...
 }
 
-const getImageData = (imageData): string => {
+const getImageData = (imageData, path): string => {
     const b64encoded = protoBase64.enc(imageData);
+    const extension = path.split(".").pop();
+    if (extension === "svg") {
+        return "data:image/svg+xml;base64," + b64encoded;
+    }
     return "data:image/png;base64," + b64encoded;
```
Comment (dae) [inline ts/image-occlusion/mask-editor.ts:129]: Ok, so we've been lying about the content type :-) While the browser tolerates this for things like .jpg images, we should probably be trying to return the correct content type for jpg/webp/gif as well.

File: ts/image-occlusion/review.ts [hunk 2]
```diff
@@ -111,6 +111,12 @@ function setupImageOcclusionInner(setupOptions?: SetupImageOcclusionOptions): vo
         } else {
             button.style.display = "none";
         }
+
+        window.addEventListener("keydown", (event) => {
+            if (event.key === "h") {
+                toggleMasks();
```
Comment (dae) [inline ts/image-occlusion/review.ts:117]: Some users use hjkl for reviewing, which this will break. Maybe we could use a capital M instead?

**Comment by dae** [comment]
Sorry for the delay here Mani - I've just gotten around to testing this. There seems to have been a regression in the positioning of images in the editor. I open the editor, select an image, and the image is offset to the right, with the right side cut off:

2023-11-13T10:47:06,814411483+10:00

If I use the zoom tool, I can move the image so it appears correctly, but when I resize the window it moves back to the right.

It appears to have regressed in d44de202943272eaebb6650991cfd844a6746c9d.

Other notes:

- Zooming seems to work well in the editor, but I would say we haven't solved #2588 until the review screen supports zooming too. Minor: when making the window bigger, the mask borders get blurry.
- The shortcuts are a nice addition, but currently they conflict with the text tool - for example, try typing a 'p' into a text area
- Very minor: rectangles and ellipses are moved inside bounds after you let go of the mouse, whereas polygons and groups are never allowed to leave the bounds. If it's very easy to make the former behave like the latter, maybe it would be worth doing that to be consistent?
- Other fixes look good!

File: ts/image-occlusion/mask-editor.ts (hunk 1 above)
Comment (krmanik) [reply ts/image-occlusion/mask-editor.ts:129]: Yeah, I have added more mime type.

File: ts/image-occlusion/review.ts (hunk 2 above)
Comment (krmanik) [reply ts/image-occlusion/review.ts:117]: It can be done.

**Comment by krmanik** [comment]
Fixing these issues added more issues, I will push the changes considering above suggested review in next commit latest by weekend. :-)

**Comment by dae** [comment]
No problem Mani, please let me know when you're ready for another review.

**Comment by dae** [comment]
Hi Mani,

There's one issue that I think we should probably solve before merging this in, as it will break the text tool for people using git otherwise. I see you've pushed https://github.com/ankitects/anki/pull/2809/commits/c68944454997664b273767095feba2de0dc2fef5, which has presumably fixed editing, but I hadn't actually noticed that issue. What I meant in my earlier comment is that when the text tool is used (below polygon), it's not possible to type text into it because the shortcuts are still active.

Re https://github.com/ankitects/anki/pull/2809/commits/5a2742766727a997f6631e7d9360408308c54263, we don't include the shortcuts inside translation strings - instead they are stored separately, so we can modify them depending on the platform, and change them without breaking translations. For example, the bold button:

```
    const keyCombination = "Control+B";
...
    <IconButton
        tooltip="{tr.editingBoldText()} ({getPlatformString(keyCombination)})"
```

**Comment by krmanik** [comment]
I have updated the shortcut implementation as suggested.

**Comment by dae** [comment]
Thanks Mani, I'll merge this in now and we can make any further tweaks in follow-up PRs.

One possible way to work around the RemoveFormatButton issue would be to only enable that shortcut when the IO mask editor is not visible. That said, I think users coming from the existing add-on may miss the existing shortcuts, that don't require the control key. Maybe an alternative solution would be to use the text:editing:entered and text:editing:exited events to set a editingTextElement flag, and only respond to those shortcut keys when the flag is not set?

**Comment by krmanik** [comment]
I will match the keyboard shortcut and use the text event in next PR.

**Pull request merged**